src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_FlowAnalysis.cs

```diff
@@ -0,0 +1,1048 @@
+                // (11,13): warning CS8602: Possible dereference of a null reference.
+                //             d[0].ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "d[0]").WithLocation(11, 13),
+                // (15,32): warning CS8619: Nullability of reference types in value of type 'C<object?>' doesn't match target type 'C<object>'.
+                //             var c = new[] { a, b };
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "b").WithArguments("C<object?>", "C<object>").WithLocation(15, 32),
+                // (17,32): warning CS8619: Nullability of reference types in value of type 'C<object>' doesn't match target type 'C<object?>'.
+                //             var d = new[] { b, a };
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "a").WithArguments("C<object>", "C<object?>").WithLocation(17, 32));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedArrayCreation_09()
+        {
+            var source =
+@"class C
+{
+    static void F(C x, Unknown? y)
+    {
+        var a = new[] { x, y };
+        a[0].ToString();
+        var b = new[] { y, x };
+        b[0].ToString();
+    }
+    static void G(C? x, Unknown y)
+    {
+        var a = new[] { x, y };
+        a[0].ToString();
+        var b = new[] { y, x };
+        b[0].ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (10,25): error CS0246: The type or namespace name 'Unknown' could not be found (are you missing a using directive or an assembly reference?)
+                //     static void G(C? x, Unknown y)
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Unknown").WithArguments("Unknown").WithLocation(10, 25),
+                // (3,24): error CS0246: The type or namespace name 'Unknown' could not be found (are you missing a using directive or an assembly reference?)
+                //     static void F(C x, Unknown? y)
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Unknown").WithArguments("Unknown").WithLocation(3, 24),
+                // (6,9): warning CS8602: Possible dereference of a null reference.
+                //         a[0].ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "a[0]").WithLocation(6, 9));
+        }
+
+        [Fact]
+        public void ExplicitlyTypedArrayCreation()
+        {
+            var source =
+@"class C
+{
+    static void F(object x, object? y)
+    {
+        var a = new object[] { x, y };
+        a[0].ToString();
+        var b = new object?[] { x, y };
+        b[0].ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,35): warning CS8601: Possible null reference assignment.
```

Comment:
It would be good to test some of those scenarios with silencing operator too.

Reply:
Added `!` cases in this test and implicitly typed array test.

---
In reply to: [157624700](https://github.com/dotnet/roslyn/pull/23439#discussion_r157624700) [](ancestors = 157624700)